### ISSUE TITLE
all connections reference to different databases of the same rds cluster connected to the same database after failover

### ISSUE THREAD

**Issue by little-apppple** [awslabs/aws-mysql-jdbc#407]
### Describe the bug

there are some problems while the we are using multi databases in the same db cluster in our application. All of the database will change to the the same one after rds failover, for expamle  when  my application configed 
`jdbc:mysql:aws://example.rds.amazonaws.com:3306/database_00?useUnicode=true&character_set_server=utf8mb4&characterEncoding=utf-8&useSSL=false` 
 `jdbc:mysql:aws://example.rds.amazonaws.com:3306/database_01?useUnicode=true&character_set_server=utf8mb4&characterEncoding=utf-8&useSSL=false`
 `jdbc:mysql:aws://example.rds.amazonaws.com:3306/database_02?useUnicode=true&character_set_server=utf8mb4&characterEncoding=utf-8&useSSL=false`
  `jdbc:mysql:aws://example.rds.amazonaws.com:3306/database_03?useUnicode=true&character_set_server=utf8mb4&characterEncoding=utf-8&useSSL=false`
  and then 
  `AuroraTopologyService.topologyCache` cached 'example.rds.amazonaws.com:3306'   originalUrl only contains jdbc:mysql:aws://example.rds.amazonaws.com:3306/database_03?useUnicode=true&character_set_server=utf8mb4&characterEncoding=utf-8&useSSL=false .

All of those connections will reference to the database of 'database_03' after failover

### Expected Behavior

all connections of this rds cluster connected to the origin database of each after failover

### Current Behavior

all connections of this rds cluster connected  to the same database after failover

### Reproduction Steps

1.config multi databases of one rds cluster in your application
2.rds cluster failover
3.write data to different databases in your application

### The AWS JDBC Driver for MySQL version used

1.1.6

### JDK version used

1.8.0_181

### Operating System and version

container: alpine-release:3.16.0

**Comment by YoungHu** [comment]
we has same issue. root cause is cluster topology is shared with clusterId and topology contains database information such as endpoint_url、db_name、username、password. in multi datasource use same cluster case, when failover recreate connection will use the topology HostInfo direct, that will cause re-conn to other database. the solution is when recreate connection, need use current connection info(db_name, username, password, db_properties) to replace topology info.
I just verify and fix `ClusterAwareWriterFailoverHandler`, hope official bug fix.

![image](https://github.com/awslabs/aws-mysql-jdbc/assets/7418602/ca324f7c-7daa-4b66-87a4-720116173f7f)
![image](https://github.com/awslabs/aws-mysql-jdbc/assets/7418602/9a633af7-1f2c-4fb9-8ce2-0ceca3125a68)

**Comment by crystall-bitquill** [comment]
Hi @little-apppple and @YoungHu, 

Thanks for reaching out and raising this issue, as well as for providing a root cause and suggested solution.

We'll take a look at this and keep you updated as we investigate.

Thank you for your patience!

**Comment by amka-ar** [comment]
We recently had the same issue and we used different `clusterId`s in URLs (`&clusterId=dbname1`) as a workaround - [to isolate caches for different URLs](https://github.com/awslabs/aws-mysql-jdbc/blob/main/src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/failover/AuroraTopologyService.java#L188).

Another observation - `initialConnectionProps` (which is added [to create mergedProps](https://github.com/awslabs/aws-mysql-jdbc/blob/main/src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/failover/ClusterAwareWriterFailoverHandler.java#L177)) doesn't have `dbname`:
```
{sessionVariables=character_set_client=utf8mb4,character_set_results=utf8mb4,character_set_connection=utf8mb4,
collation_connection=utf8mb4_unicode_ci, logger=Slf4JLogger, elideSetAutoCommits=true, 
failureDetectionInterval=2000, failureDetectionCount=1, failureDetectionTime=5000, useLocalSessionState=true}
```
but `HostInfo` from the topology cache have it (from the first connection made):
```
{logger=Slf4JLogger, failureDetectionCount=1, sessionVariables=character_set_client=utf8mb4,
character_set_results=utf8mb4, character_set_connection=utf8mb4,collation_connection=utf8mb4_unicode_ci, 
TOPOLOGY_SERVICE_SERVER_ID=ottodb-2022-nov-cluster-0, dbname=core, 
TOPOLOGY_SERVICE_SESSION_ID=MASTER_SESSION_ID, elideSetAutoCommits=true, connectTimeout=0, socketTimeout=0, 
TOPOLOGY_SERVICE_LAST_UPDATE_TIMESTAMP=2023-05-16 18:49:59.22666, failureDetectionInterval=2000, 
TOPOLOGY_SERVICE_REPLICA_LAG_IN_MILLISECONDS=0.0, failureDetectionTime=5000, useLocalSessionState=true}
```
it is added [here](https://github.com/awslabs/aws-mysql-jdbc/blob/main/src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/failover/FailoverConnectionPlugin.java#L828). As a result of `hostInfo.hostProperties + initialConnectionProps`, connection for the second URL (which uses another schema, not "core") will get `dbname=core`.

It might be safer to have separate caches for unique URLs since `url1` might have some extra parameters defined compared to `url2` and theses parameters will be copied to `url2` connection

**Comment by crystall-bitquill** [comment]
Hi @amka-ar,

Thank you for providing your observations and suggestions!

**Comment by crystall-bitquill** [comment]
Hi @little-apppple, @YoungHu, and @amka-ar,

We've just merged a fix (#410). Could you kindly check out our [snapshot build](https://github.com/awslabs/aws-mysql-jdbc#using-a-snapshot-of-the-driver) and let us know if the issue persists?

Thank you!

**Comment by crystall-bitquill** [comment]
Hi @little-apppple, @YoungHu, and @amka-ar,

I wanted to check in to see if the issue was resolved with the snapshot build.

If there are no more updates on this ticket in the next few days, it will be closed. However, if there are any further questions or comments please feel free to create a new issue or to reopen this one.

Thank you!

**Comment by YoungHu** [comment]
> Hi @little-apppple, @YoungHu, and @amka-ar,
> 
> I wanted to check in to see if the issue was resolved with the snapshot build.
> 
> If there are no more updates on this ticket in the next few days, it will be closed. However, if there are any further questions or comments please feel free to create a new issue or to reopen this one.
> 
> Thank you!

found new issue, debugging!!!

**Comment by YoungHu** [comment]
got metadata pattern check failed error.
```log
### Error updating database.  Cause: org.apache.shardingsphere.underlying.common.database.metadata.UnrecognizedDatabaseURLException: The URL: 'jdbc:mysql:aws://instance-1-name-southeast-1.rds.amazonaws.com:3306/' is not recognized. Please refer to this pattern: 'jdbc:(mysql|mysqlx)(:loadbalance|:replication)?:(\w*:)?//([\w\-\.]+):?([0-9]*),?.*?/([\w\-]+);?\S*'.
```

**Comment by crystall-bitquill** [comment]
Hi @YoungHu, thanks for raising this issue!

To clarify, are you using Sharding JDBC Core? My dependencies include this: `implementation("org.apache.shardingsphere:sharding-jdbc-core:4.1.1")`.

I tried to reproduce this using the following code:

```java
MysqlDataSource dataSource1 = new MysqlDataSource();
    dataSource1.setUrl("jdbc:mysql:aws://db-identifier.cluster-XYZ.us-east-2.rds.amazonaws.com/database1?logger=StandardLogger");
    dataSource1.setUser("user");
    dataSource1.setPassword("password");

MysqlDataSource dataSource2 = new MysqlDataSource();
    dataSource2.setUrl("jdbc:mysql:aws://db-identifier.cluster-XYZ.us-east-2.rds.amazonaws.com/database2?logger=StandardLogger");
    dataSource2.setUser("user");
    dataSource2.setPassword("password");

Map<String, DataSource> dataSourceMap = new HashMap<>();
dataSourceMap.put("ds_1", dataSource1);
dataSourceMap.put("ds_2", dataSource2);

ShardingRuleConfiguration ruleConfiguration = new ShardingRuleConfiguration();
DataSource dataSource = ShardingDataSourceFactory.createDataSource(dataSourceMap, ruleConfiguration, new Properties());

try (final Connection conn = dataSource.getConnection()) {
    // execute queries
}
```

However, I wasn't able to get the same error. Is there anything in your setup that is different or might be missing from the sample above? Or would you perhaps be able to share a snippet of the code used when you encountered the problem?

**Comment by YoungHu** [comment]
> Hi @YoungHu, thanks for raising this issue!
> 
> To clarify, are you using Sharding JDBC Core? My dependencies include this: `implementation("org.apache.shardingsphere:sharding-jdbc-core:4.1.1")`.
> 
> I tried to reproduce this using the following code:
> 
> ```java
> MysqlDataSource dataSource1 = new MysqlDataSource();
>     dataSource1.setUrl("jdbc:mysql:aws://db-identifier.cluster-XYZ.us-east-2.rds.amazonaws.com/database1?logger=StandardLogger");
>     dataSource1.setUser("user");
>     dataSource1.setPassword("password");
> 
> MysqlDataSource dataSource2 = new MysqlDataSource();
>     dataSource2.setUrl("jdbc:mysql:aws://db-identifier.cluster-XYZ.us-east-2.rds.amazonaws.com/database2?logger=StandardLogger");
>     dataSource2.setUser("user");
>     dataSource2.setPassword("password");
> 
> Map<String, DataSource> dataSourceMap = new HashMap<>();
> dataSourceMap.put("ds_1", dataSource1);
> dataSourceMap.put("ds_2", dataSource2);
> 
> ShardingRuleConfiguration ruleConfiguration = new ShardingRuleConfiguration();
> DataSource dataSource = ShardingDataSourceFactory.createDataSource(dataSourceMap, ruleConfiguration, new Properties());
> 
> try (final Connection conn = dataSource.getConnection()) {
>     // execute queries
> }
> ```
> 
> However, I wasn't able to get the same error. Is there anything in your setup that is different or might be missing from the sample above? Or would you perhaps be able to share a snippet of the code used when you encountered the problem?

yes, it use `org.apache.shardingsphere:sharding-jdbc-core:4.1.1`, when debugging, sharding-jdbc call datasource metadata.getUrl, this method implement by aws mysql jdbc driver which removed db properties, so the url just return host info, missing the properties.

**Comment by crystall-bitquill** [comment]
Hi @YoungHu,

A fix has been merged (#421), could you kindly check out the [snapshot](https://github.com/awslabs/aws-mysql-jdbc#using-a-snapshot-of-the-driver) build again and see if that resolves the issue?

Thank you!

**Comment by YoungHu** [comment]
> Hi @YoungHu,
> 
> A fix has been merged (#421), could you kindly check out the [snapshot](https://github.com/awslabs/aws-mysql-jdbc#using-a-snapshot-of-the-driver) build again and see if that resolves the issue?
> 
> Thank you!

is ok now.

**Comment by karenc-bq** [comment]
Thank you @YoungHu, we will be closing the ticket now.
